derive.h

```diff
@@ -45,28 +49,32 @@ ostream& operator<<(ostream& os, const Expression& ex) {
 
 class CoefExpression : public Expression {
 public:
-	CoefExpression(double a_coef): coef(a_coef) 
+	CoefExpression(double coef): m_coef(coef) 
 	{ }
 	
-	virtual bool is_zero() { return coef == 0.; }
+	virtual bool is_zero() { return coef() == 0.; }
 	virtual shared_ptr<Expression> derivative() const {
 		return std::dynamic_pointer_cast<Expression>(
 			shared_ptr<CoefExpression>(new CoefExpression{0})
 		);
 	}
 
 	virtual ostream& printTo(ostream& os) const {
-		os << coef;
+		os << coef();
 		return os;
 	}
-	double coef;
+	double coef() const { return m_coef;}
+	friend class BinaryExpression;
+protected:
+	virtual optional<char> opt_var() const { return optional<char>{}; }
```

Comment:
We have `std::nullopt` for this purpose.

Reply:
Ok, great, fixed.